### PR TITLE
Speedup iteration for wrappers around arrays

### LINKED ISSUE #39354: Regression for iteration over Vector{Int} wrapper types
Iteration over thin wrappers of certain `Vector{T}`s seems to have regressed on 1.6.0-beta1.0 vs. 1.5.1:

By way of example: defining a thin wrapper over a `Vector{T}`:
```jl
import Base: size, getindex, IndexStyle

struct V{T} <: AbstractVector{T} # wrapper type over a vector
    x::Vector{T}
end
size(v::V) = size(v.x)
Base.@propagate_inbounds getindex(v::V, i::Int) = v.x[i]
IndexStyle(::Type{<:V{T}}) where T = IndexStyle(Vector{T}) # ... IndexLinear()
```
and then doing a simple `sum` implementation for that:
```jl
function f(v)
    s = zero(eltype(v))
    for vᵢ in v
        s += vᵢ
    end
    return s
end
```

I see a performance regression on 1.6.0-beta1 vs. 1.5.1 when the wrapped vector is a `Vector{Int}` (but not when it is a `Vector{Float64}`):

**On 1.5.1:**
```jl
using BenchmarkTools
v_int   = V(rand(1:10, 100000)) # wrapped Vector{Int}
v_float = V(rand(100000))       # wrapped Vector{Float64}

@btime f($v_int)   # 40.499 μs
@btime f($v_float) # 117.999 μs
```

**On 1.6.0-beta1.0**
```jl
@btime f($v_int)   # 88.500 μs <-- 2x regression
@btime f($v_float) # 117.999 μs
```

There's a possibly related issue (albeit probably not related to the regression per se) where it is significantly faster to iterate over the wrapped vector itself when it is a `Vector{Int}`, as I noted in this [Discourse post](https://discourse.julialang.org/t/iteration-getindex-performance-of-abstractarray-wrapper-types/53729).

### PR DESCRIPTION
**Pull request by p-zubieta**
This will make `for` loops such as `for i in itr; #= body =#; end` lower to

```julia
@inbounds next = iterate(iter)
while next !== nothing
    (i, state) = next
    # body
    @inbounds next = iterate(iter, state)
end
```

Currently no `@inbounds` annotations are added unless the whole for loop is manually annotated. This should make, for instance, iteration over `views` faster.

 Before
```julia
using BenchmarkTools

let
    v = rand(100);
    w = view(v, :)
    @btime any(iszero, $v)  # 41.471 ns (0 allocations: 0 bytes)
    @btime any(iszero, $w)  # 61.306 ns (0 allocations: 0 bytes)
end
```

After
```julia
using BenchmarkTools

let
    v = rand(100);
    w = view(v, :)
    @btime any(iszero, $v)  # 41.471 ns (0 allocations: 0 bytes)
    @btime any(iszero, $w)  # 41.840 ns (0 allocations: 0 bytes)
end
```

Also fixes #39354.

### CODE REVIEW

File: base/abstractarray.jl [hunk 1]
```diff
@@ -1091,6 +1091,7 @@ zero(x::AbstractArray{T}) where {T} = fill!(similar(x), zero(T))
 # own, IndexCartesian's CartesianIndices is more complicated and requires explicit
 # inlining.
 function iterate(A::AbstractArray, state=(eachindex(A),))
+    @_propagate_inbounds_meta
     y = iterate(state...)
     y === nothing && return nothing
     A[y[1]], (state[1], tail(y)...)
```
Comment (vtjnash) [inline base/abstractarray.jl:1097]: I thought it was `A[I]` that should be always inbounds here, always

Comment (pabloferz) [reply base/abstractarray.jl:1097]: I thought that too, but given the kwarg `state`, I don't know if there is a chance for someone to pass one that would make `A[y[1]]` out of bounds.

Should I just mark this as `@inbounds` and remove the rest of the changes?

**Comment by JeffBezanson** [comment]
I believe we've considered this way too dangerous in the past. It basically makes part of the language unsafe by default.

**Comment by pabloferz** [comment]
Oh right. Any `for` loop would lower `iterate` but if iterate is marked incorrectly as `@inbounds` things will, break. Will have to think of something else.

**Comment by vtjnash** [comment]
Triage agrees this is too dangerous for a default

**Pull request closed without merging**